schematic/store/synapse_tracker.py

```diff
@@ -0,0 +1,140 @@
+"""This script is responsible for creating a 'pull through cache' class that can be
+added through composition to any class where Synapse entities might be used. The idea
+behind this class is to provide a mechanism such that if a Synapse entity is requested
+multiple times, the entity is only downloaded once. This is useful for preventing
+multiple downloads of the same entity, which can be time consuming."""
+from dataclasses import dataclass, field
+from typing import Dict, List, Union
+
+import synapseclient
+from synapseclient import Entity, File, Folder, Project, Schema
+
+
+@dataclass
+class SynapseEntiyTracker:
+    """The SynapseEntiyTracker class handles tracking synapse entities throughout the
+    lifecycle of a request to schematic. It is used to prevent multiple downloads of
+    the same entity."""
+
+    synapse_entities: Dict[str, Union[Entity, Project, File, Folder, Schema]] = field(
+        default_factory=dict
+    )
+    project_headers: Dict[str, List[Dict[str, str]]] = field(default_factory=dict)
+    """A dictionary of project headers for each user requested."""
+
+    def get(
+        self,
+        synapse_id: str,
+        syn: synapseclient.Synapse,
+        download_file: bool = False,
+        retrieve_if_not_present: bool = True,
+        download_location: str = None,
+        if_collision: str = None,
+    ) -> Union[Entity, Project, File, Folder, Schema]:
+        """Retrieves a Synapse entity from the cache if it exists, otherwise downloads
+        the entity from Synapse and adds it to the cache.
+
+        Args:
+            synapse_id: The Synapse ID of the entity to retrieve.
+            syn: A Synapse object.
+            download_file: If True, download the file.
+            retrieve_if_not_present: If True, retrieve the entity if it is not present in the cache.
+            download_location: The location to download the file to.
+            if_collision: The action to take if there is a collision when downloading the file.
+
+        Returns:
+            The Synapse entity.
+        """
+        entity = self.synapse_entities.get(synapse_id, None)
+
+        if entity is None or (download_file and not entity.path):
+            if not retrieve_if_not_present:
+                return None
+            entity = syn.get(
+                synapse_id,
+                downloadFile=download_file,
+                downloadLocation=download_location,
+                ifcollision=if_collision,
+            )
+        self.synapse_entities.update({synapse_id: entity})
+        return entity
+
+    def add(
+        self, synapse_id: str, entity: Union[Entity, Project, File, Folder, Schema]
+    ) -> None:
+        """Adds a Synapse entity to the cache.
+
+        Args:
+            synapse_id: The Synapse ID of the entity to add.
+            entity: The Synapse entity to add.
+        """
+        self.synapse_entities.update({synapse_id: entity})
+
+    def remove(self, synapse_id: str) -> None:
```

Comment:
Thanks for adding this @BryanFauble ! I am curious what do you envision the use case for this?

Reply:
See https://sagebionetworks.jira.com/browse/PLFM-8605 - We must remove entities from the local cache when the table data is updated.

Reply:
Hi @BryanFauble ! Thanks for providing more context. I read the Jira ticket that you linked but I don't think that I understand. Based on the ticket and the comment, I am guessing that `store` function is not returning the etag of table because UploadToTableResult.etag is not the etag of the table. But why is it related to removing entities from the local cache?

Reply:
@linglp Let's recap the concept of eTag and what it's used for in Synapse:

> Synapse employs an Optimistic Concurrency Control (OCC) scheme to handle concurrent updates. Since the E-Tag changes every time an entity is updated it is used to detect when a client's current representation of an entity is out-of-date.

We can take a simple example where suppose we create a Table and it's assigned an etag of `1`, I go ahead and add an annotation to the table and that Table now has a new eTag `2`. When I updated those annotations we will want to make sure that we capture in the local cache:

1. The current state of the entity (This includes all of the traditional fields about the Entity, AND, the annotations for that entity)
2. The updated eTag as a result of saving those annotations.

The above works because when we make the change to add those annotations the Synapse API gives us back the eTag of the updated entity and we can store that back into our local in-memory cache.


In the case of updates to the data or schema of a table the eTag is also getting updated. Let's suppose we add data to a table, Synapse will now update that eTag to `3`, however, us as the caller have no way of knowing that Synapse updated the eTag to `3`, we only ever know it was `2`, and we never got the updated eTag as a result of submitting our transaction. 

Ask: What would occur if I try to submit another transaction on that Table and pass along my known eTag of `2`? Synapse would reject the transaction because we on the client side do not have the most up to date version of that entity that Synapse has. This is the root of what the ticket is talking about/addressing.


Why do we need to remove the entity from our local in-memory cache? It's the result of the `Ask` above. Since we do not have the most up to date eTag of that entity, the next time we read for that entity in order to perform some operation, that read MUST be sent to Synapse. We cannot trust our local in-memory cache anymore because we know we do not have the up to date eTag of the entity. That is what the `.remove()` operation is for.